pkg/kv/kvserver/scheduler.go

```diff
@@ -185,38 +165,63 @@ var raftSchedulerBatchPool = sync.Pool{
 
 // raftSchedulerBatch is a batch of range IDs to enqueue. It enables
 // efficient per-shard enqueueing.
-type raftSchedulerBatch [][]roachpb.RangeID // by shard
+type raftSchedulerBatch struct {
+	rangeIDs    [][]roachpb.RangeID // by shard
+	priorityIDs map[roachpb.RangeID]bool
+}
 
-func newRaftSchedulerBatch(numShards int) raftSchedulerBatch {
+func newRaftSchedulerBatch(numShards int, priorityIDs *syncutil.IntMap) *raftSchedulerBatch {
 	b := raftSchedulerBatchPool.Get().(*raftSchedulerBatch)
-	if len(*b) != numShards {
-		*b = make([][]roachpb.RangeID, numShards)
+	if len(b.rangeIDs) != numShards {
+		b.rangeIDs = make([][]roachpb.RangeID, numShards)
+	}
+	if b.priorityIDs == nil {
+		b.priorityIDs = make(map[roachpb.RangeID]bool, 8) // expect few ranges, if any
 	}
-	return *b
+	// Cache the priority range IDs in an owned map, since we expect this to be
+	// very small or empty and we do a lookup for every Add() call.
+	priorityIDs.Range(func(id int64, _ unsafe.Pointer) bool {
+		b.priorityIDs[roachpb.RangeID(id)] = true
+		return true
+	})
+	return b
 }
 
-func (b raftSchedulerBatch) Add(id roachpb.RangeID) {
-	shard := int(id) % len(b)
-	b[shard] = append(b[shard], id)
+func (b *raftSchedulerBatch) Add(id roachpb.RangeID) {
+	shardIdx := shardIndex(id, len(b.rangeIDs), b.priorityIDs[id])
+	b.rangeIDs[shardIdx] = append(b.rangeIDs[shardIdx], id)
 }
 
-func (b raftSchedulerBatch) Reset() {
-	for i := range b {
-		b[i] = b[i][:0]
+func (b *raftSchedulerBatch) Close() {
+	for i := range b.rangeIDs {
+		b.rangeIDs[i] = b.rangeIDs[i][:0]
+	}
+	for i := range b.priorityIDs {
+		delete(b.priorityIDs, i)
 	}
+	raftSchedulerBatchPool.Put(b)
 }
 
-func (b raftSchedulerBatch) Close() {
-	b.Reset()
-	raftSchedulerBatchPool.Put(&b)
+// shardIndex returns the raftScheduler shard index of the given range ID based
+// on the shard count and the range's priority. Priority ranges are assigned to
+// the reserved shard 0, other ranges are modulo range ID (ignoring shard 0).
+func shardIndex(id roachpb.RangeID, numShards int, priority bool) int {
+	if priority {
+		return 0
+	}
+	return 1 + int(int64(id)%int64(numShards-1)) // int64s to avoid overflow
```

Comment:
Just checking: This assumes that `numShards > 1`, right?
I.e. `numShards == 1 /* priotity */ + numRegularShards /* non-priority */`

Reply:
Yep, there will always be one priority shard and one regular shard. Added a comment.